openedx/core/djangoapps/course_roles/docs/decisions/0002-data_storage.rst

```diff
@@ -0,0 +1,56 @@
+2. Data Storage
+################
+
+Status
+******
+
+**Provisional** *2023-09-13*
+
+Context
+*******
+
+The requirements of the course roles project indicate that equal priority be given to user understanding of the roles and flexibility of the system. It has been determined that in order to prioritize user understadning of roles and permissions it is necessary that the descriptions be localized.
+
+**Localization:** It is necessary for all text that will be displayed in the LMS to be shown in the appropriate language. The current process for translation requires that all strings displayed to users be in strings in the codebase.
+
+**Flexibility:** It is necessary that course_roles allows for the easy creation of new roles. It is considered important, but not strictly required, that each instance of the Open edX code can have its own roles without needing to fork the repo and maintain separate code. 
+
+Decision
+********
+
+We will be using a combination of data storage in the database and data storage structures within the code.
+We will store the roles, permissions, role permissions (rolepermission table), and role assignments (userrole table) in the database. 
+We will store the role names (for default roles), permission names, and permission descriptions in a data object in the code.
+We will write code that pulls the role name from the database table for any role not found in the data object in the code. This applies to non-default roles that may differ from instance to instance.
+We will update both the code and the database table rows if a new permission is added.
+We will update both the code and the database table rows if a new default row is added.
+We will only add data in the database if a role is being added to a single Open edX instance.
+
+Consequences
+************
+
+This decision will mean that any futurue default role additions or permissions changes will require changes in the code and the database. It also means that there is a chance of a default role name being listed in the UI using the name value in the database. This would occur if a role was added to the database, but the role was not added to the data structure in the code.
```

Comment:
```suggestion
This decision will mean that any future default role additions or permissions changes will require changes in the code and the database. It also means that there is a chance of a default role name being listed in the UI using the name value in the database. This would occur if a role was added to the database, but the role was not added to the data structure in the code.
```

Reply:
So If I create a new role, I'll need to open a PR updating the data structure in the code? how would that look like for custom specific roles?

Reply:
If a new role is being created that should be available for use by all instances, it would require a PR with a migration and additional code in the data structure.

If a new role is being created for a single instance, it can be done by adding the role to the db. If it is a single instance role, but needs to be available in multiple languages, it can be added to the db in each required language with the roles linked to the same permissions.

Please note, this decision is not fully implemented with this PR.